File: app/har.py
"""Reqable 上报服务器 HAR 解析(纯函数,便于单测)。

Reqable 的「上报服务器」功能把每个已完成的 HTTP 会话按 HAR 格式(JSON)
POST 到用户自建服务端,可选 gzip / brotli / zstd 压缩(通过 Content-Encoding 标识)。
本模块负责:解压请求体 -> 解析 HAR JSON -> 从会话中提取原始响应/请求字节。
"""
import base64
import gzip
import json
from typing import Iterator, Optional

# HAR 中二进制内容(响应 content / 请求 postData)的 base64 标识
ENCODING_BASE64 = "base64"


def decompress_body(raw: bytes, content_encoding: Optional[str] = None) -> bytes:
    """按 Content-Encoding 解压请求体;无编码或 identity 时原样返回。"""
    enc = (content_encoding or "").strip().lower()
    if not enc or enc in ("identity",):
        return raw
    if enc in ("gzip", "x-gzip"):
        return gzip.decompress(raw)
    if enc == "br":
        import brotli

        return brotli.decompress(raw)
    if enc in ("zstd", "zstandard"):
        import zstandard

        # 注意:Reqable(流式压缩)产出的 zstd 帧头通常不携带内容大小,
        # ZstdDecompressor().decompress() 会报
        # "could not determine content size in frame header",
        # 因此必须用不依赖帧头 content size 的 decompressobj()。
        return zstandard.ZstdDecompressor().decompressobj().decompress(raw)
    raise ValueError(f"Unsupported Content-Encoding: {content_encoding!r}")


def parse_har(raw: bytes, content_encoding: Optional[str] = None) -> dict:
    """解压并解析 HAR JSON。

    解压失败时兜底按未压缩 JSON 直接解析(Reqable 偶发会带编码头却发未压缩体,
    失败不重试,尽量宽容)。
    """
    try:
        body = decompress_body(raw, content_encoding)
    except Exception:
        body = raw
    return json.loads(body.decode("utf-8"))


def har_entries(har: dict) -> list:
    """取出 HAR 中的所有会话条目(log.entries)。"""
    return (har.get("log") or {}).get("entries") or []


def content_to_bytes(obj: Optional[dict]) -> Optional[bytes]:
    """把 HAR 的 response.content / request.postData 对象还原为原始字节。

    HAR 规范约定:二进制正文存于 text 字段并附 encoding="base64",
    此时 text 为 base64 编码;无 encoding 时按 UTF-8 文本处理。
    """
    if not obj:
        return None
    text = obj.get("text")
    if text is None:
        return None
    if isinstance(text, bytes):
        return text
    if not isinstance(text, str):
        return None
    if obj.get("encoding") == ENCODING_BASE64:
        return base64.b64decode(text)
    return text.encode("utf-8", errors="replace")


def entry_candidate_bodies(entry: dict) -> Iterator[bytes]:
    """按优先级产出某个会话里可能携带存档的原始字节。

    优先响应体(MySekai 数据在游戏 API 的响应中),其次请求体。
    """
    response = entry.get("response") or {}
    body = content_to_bytes(response.get("content"))
    if body:
        yield body
    request = entry.get("request") or {}
    body = content_to_bytes(request.get("postData"))
    if body:
        yield body

File: internal/har/har.go
// Package har parses Reqable HAR reports without retaining raw report bodies.
package har

import (
	"bytes"
	"compress/gzip"
	"encoding/base64"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"strings"

	"github.com/andybalholm/brotli"
	"github.com/klauspost/compress/zstd"
)

var (
	ErrBodyTooLarge               = errors.New("HAR body exceeds configured limit")
	ErrUnsupportedContentEncoding = errors.New("unsupported Content-Encoding")
)

// DecompressBody decodes a request body according to Content-Encoding and
// bounds both compressed identity input and decompressed output.
func DecompressBody(raw []byte, contentEncoding string, maxSize int64) ([]byte, error) {
	if maxSize <= 0 {
		return nil, fmt.Errorf("invalid maximum body size %d", maxSize)
	}
	encoding := strings.ToLower(strings.TrimSpace(contentEncoding))
	switch encoding {
	case "", "identity":
		if int64(len(raw)) > maxSize {
			return nil, ErrBodyTooLarge
		}
		return raw, nil
	case "gzip", "x-gzip":
		reader, err := gzip.NewReader(bytes.NewReader(raw))
		if err != nil {
			return nil, err
		}
		defer reader.Close()
		return readLimited(reader, maxSize)
	case "br":
		return readLimited(brotli.NewReader(bytes.NewReader(raw)), maxSize)
	case "zstd", "zstandard":
		reader, err := zstd.NewReader(bytes.NewReader(raw), zstd.WithDecoderConcurrency(1))
		if err != nil {
			return nil, err
		}
		defer reader.Close()
		return readLimited(reader, maxSize)
	default:
		return nil, fmt.Errorf("%w: %q", ErrUnsupportedContentEncoding, contentEncoding)
	}
}

func readLimited(reader io.Reader, maxSize int64) ([]byte, error) {
	body, err := io.ReadAll(io.LimitReader(reader, maxSize+1))
	if err != nil {
		return nil, err
	}
	if int64(len(body)) > maxSize {
		return nil, ErrBodyTooLarge
	}
	return body, nil
}

// Parse decodes a HAR JSON object. Reqable occasionally sends an uncompressed
// JSON body with a compression header; when decompression fails, this preserves
// a JSON-only fallback for that malformed-client behavior.
func Parse(raw []byte, contentEncoding string, maxSize int64) (map[string]any, error) {
	body, err := DecompressBody(raw, contentEncoding, maxSize)
	if err != nil {
		trimmed := bytes.TrimSpace(raw)
		if !bytes.HasPrefix(trimmed, []byte("{")) && !bytes.HasPrefix(trimmed, []byte("[")) {
			return nil, err
		}
		if int64(len(raw)) > maxSize {
			return nil, ErrBodyTooLarge
		}
		body = raw
	}
	var document map[string]any
	if err := json.Unmarshal(body, &document); err != nil {
		return nil, fmt.Errorf("decode HAR JSON: %w", err)
	}
	return document, nil
}

// Entries returns log.entries, treating malformed or missing data as empty.
func Entries(document map[string]any) []map[string]any {
	log, ok := object(document["log"])
	if !ok {
		return nil
	}
	values, ok := array(log["entries"])
	if !ok {
		return nil
	}
	entries := make([]map[string]any, 0, len(values))
	for _, value := range values {
		if entry, ok := object(value); ok {
			entries = append(entries, entry)
		}
	}
	return entries
}

// CandidateBodies returns non-empty response content before request postData,
// matching the MySekai response-first extraction rule.
func CandidateBodies(entry map[string]any) ([][]byte, error) {
	candidates := make([][]byte, 0, 2)
	if response, ok := object(entry["response"]); ok {
		body, present, err := ContentToBytes(response["content"])
		if err != nil {
			return nil, err
		}
		if present && len(body) > 0 {
			candidates = append(candidates, body)
		}
	}
	if request, ok := object(entry["request"]); ok {
		body, present, err := ContentToBytes(request["postData"])
		if err != nil {
			return nil, err
		}
		if present && len(body) > 0 {
			candidates = append(candidates, body)
		}
	}
	return candidates, nil
}

// ContentToBytes restores a HAR response.content or request.postData body.
func ContentToBytes(value any) ([]byte, bool, error) {
	content, ok := object(value)
	if !ok {
		return nil, false, nil
	}
	text, exists := content["text"]
	if !exists || text == nil {
		return nil, false, nil
	}
	var body []byte
	switch typed := text.(type) {
	case string:
		body = []byte(typed)
	case []byte:
		body = typed
	default:
		return nil, false, nil
	}
	encoding, _ := content["encoding"].(string)
	if encoding == "base64" {
		decoded, err := base64.StdEncoding.DecodeString(string(body))
		if err != nil {
			return nil, false, fmt.Errorf("decode HAR base64 content: %w", err)
		}
		return decoded, true, nil
	}
	return body, true, nil
}

// RequestURL extracts a HAR entry request URL or returns an empty string.
func RequestURL(entry map[string]any) string {
	request, ok := object(entry["request"])
	if !ok {
		return ""
	}
	url, _ := request["url"].(string)
	return url
}

func object(value any) (map[string]any, bool) {
	result, ok := value.(map[string]any)
	return result, ok
}

func array(value any) ([]any, bool) {
	result, ok := value.([]any)
	return result, ok
}
